lib/core/re.nit

```diff
@@ -367,31 +367,32 @@ class Regex
 		assert native != null
 
 		# Actually execute
-		text = text.to_s
-		var sub = text.substring_from(from)
-		var cstr = sub.to_cstring
-		var bstr = new FlatString.full(cstr, sub.bytelen, 0, sub.bytelen - 1, text.length - from)
+		var cstr = text.to_cstring
+		var rets = cstr.to_s_with_length(text.bytelen)
+		var bytefrom = cstr.char_to_byte_index_cached(charfrom, 0, 0)
+		var subcstr = cstr.fast_cstring(bytefrom)
 		var eflags = gather_eflags
 		var native_match = self.native_match
 
 		var nsub = native.re_nsub
-		var res = native.regexec(cstr, nsub+1, native_match, eflags)
+		var res = native.regexec(subcstr, nsub + 1, native_match, eflags)
 
 		# Found one?
 		if res == 0 then
-			var first_char = bstr.byte_to_char_index(native_match.rm_so)
-			var length_char = bstr.byte_to_char_index(native_match.rm_eo - native_match.rm_so - 1) # FIXME For issue #1684
-			var match = new Match(text,
-				from + first_char,
-				length_char + 1)
+			var bfrom = native_match.rm_so + bytefrom
+			var bto = native_match.rm_eo - 1 + bytefrom
+			var cpos = cstr.byte_to_char_index_cached(bfrom, charfrom, bytefrom)
+			var len = cstr.utf8_length(bfrom, bto)
+			var match = new Match(rets, cpos, len)
 
 			# Add sub expressions
 			for i in [1 .. nsub] do
-				first_char = bstr.byte_to_char_index(native_match[i].rm_so)
-				length_char = bstr.byte_to_char_index(native_match[i].rm_eo - native_match[i].rm_so - 1) # FIXME For issue #1684
-				match.subs.add new Match( text,
-					from + first_char,
-					length_char + 1)
+				if native_match[i].rm_so < 0 then continue
```

Comment:
I don't know about this, I'm not familiar with the C API, though I suppose that if so is -1, this is a no match.

In our case, I suppose it would not be that much of an issue to have sub matches as an array of nullable Match

I can change the API, but the actual behaviour is in my opinion acceptable, if no submatch is encountered, the array will be empty and I think I'd be fine with that.

Although the correspondence between the sub expression is lost, does it really matter which sub expression is matched to a substring ?
I guess there are some cases in which it might though I can't see any.